Add preflight OS, CPU, RAM, Swap, and Filesystem checks

Kushal-deb opened the pull request:
- Implemented OS, NIC and Other preflight checks to validate system requirements before Ceph cluster creation.
    
    - Checks include:
      - OS version (RHEL 9+ required)
      - SELinux enforcing mode
      - Firewalld installation and status
      - Required package availability (rpcbind, podman, firewalld)
      - Podman version check (>= 3.3)
      - RHEL software profile validation
      - Tuned profile check
      - CPU, RAM, Swap, and Filesystem (part of other checks)
      - Check whether jumbo frames are enabled
      - Is it configured with DHCP or static IP
      - Is the bandwidth sufficient
      - Collect and output current NIC options set (e.g. Bonding, not bridged or virtual)
      - Check and report network latency (ping) with all hosts provided in the inventory file
      - Separate NICs for front-end and back-end networks
  
Enhancements: 
- https://tracker.ceph.com/issues/69726
- https://tracker.ceph.com/issues/69781
- https://tracker.ceph.com/issues/69843

  
  
  Logs:
  
```
❯ ansible-playbook -i ~/ansible-inventory/inventory.ini cephadm-preflight.yml                                                                                                                                                              ─╯

PLAY [insecure_registries] *******************************************************************************************************************************************************************************************************************

TASK [fail if insecure_registry is undefined] ************************************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

PLAY [preflight] *****************************************************************************************************************************************************************************************************************************

TASK [fail when ceph_origin is custom with no repository defined] ****************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

TASK [fail if baseurl is not defined for ceph_custom_repositories] ***************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

PLAY [all] ***********************************************************************************************************************************************************************************************************************************

❯ ansible-playbook -i ~/ansible-inventory/inventory.ini cephadm-preflight.yml                                                                                                                                                              ─╯

PLAY [insecure_registries] *******************************************************************************************************************************************************************************************************************

TASK [fail if insecure_registry is undefined] ************************************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

PLAY [preflight] *****************************************************************************************************************************************************************************************************************************

TASK [fail when ceph_origin is custom with no repository defined] ****************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

TASK [fail if baseurl is not defined for ceph_custom_repositories] ***************************************************************************************************************************************************************************
skipping: [rhel-ceph-admin]

PLAY [Preflight Checks for Ceph Deployment] **************************************************************************************************************************************************************************************************

TASK [Initialize preflight results list] *****************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Collect installed package facts] *******************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Check if OS is RHEL 9+] ****************************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Ensure SELinux is set to Enforcing mode] ***********************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Determine SELinux Check Result] ********************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Determine SELinux Failure Reason] ******************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Determine Package Installation Check Result] *******************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Determine Package Installation Failure Reason] *****************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Fetch Firewalld status] ****************************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Extract Podman version if installed] ***************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Determine if Podman meets version requirement (>=3.3)] *********************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Validate RHEL software profile] ********************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Define RHEL Profile Check Result] ******************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Define RHEL Profile Check Reason] ******************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Get current tuned profile] *************************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Define Tuned Profile Check Result] *****************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Define Tuned Profile Check Reason] *****************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Check CPU x86-64-v2 support] ***********************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Define CPU, RAM, Swap, and Filesystem Check Variables] *********************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Ping all hosts in inventory to measure latency] ****************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin] => (item=rhel-ceph-admin)

TASK [Define networking facts] ***************************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Store all preflight check results] *****************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Generate preflight check report file] **************************************************************************************************************************************************************************************************
changed: [rhel-ceph-admin -> localhost]

TASK [Load the preflight check report] *******************************************************************************************************************************************************************************************************
ok: [rhel-ceph-admin]

TASK [Final Check - Fail if any critical checks failed] **************************************************************************************************************************************************************************************
fatal: [rhel-ceph-admin]: FAILED! => changed=false 
  msg: 'Preflight checks failed for the following: Tuned Profile, RHEL Profile, Minimum RAM, Swap Space, /var Partition, Root Filesystem, Jumbo Frames Enabled, NIC Static IP Configuration, NIC Bandwidth. Please resolve these issues before proceeding.'

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
rhel-ceph-admin            : ok=25   changed=1    unreachable=0    failed=1    skipped=3    rescued=0    ignored=0  

=================================================================================================

❯ cat preflight_report.txt                                                                                                                                                                                                                 ─╯
==================================================
               **  Preflight Check Report **
==================================================

 System Checks
--------------------------------------------------
- OS Version: ✅ Passed

- Tuned Profile: ❌ Failed
    - Reason: Incorrect tuned profile. Expected: throughput-performance

- RHEL Profile: ❌ Failed
    - Reason: Incorrect RHEL software profile. Expected: Server with File and Storage Server.

- Firewalld Running: ✅ Passed

- Podman Installed: ✅ Passed

- SELinux: ✅ Passed

- Required Packages Installed: ✅ Passed

- Minimum RAM (8GB): ❌ Failed
    - Reason: System has only 7684 MB RAM, required: 8192MB

- Swap Space (1.5x RAM): ❌ Failed
    - Reason: System has only 5119 MB Swap, required: 11526 MB

- CPU x86-64-v2: ✅ Passed

- CPU Cores >= 4: ✅ Passed

- /var is a separate partition: ❌ Failed
    - Reason: /var is not a separate partition

- Root Filesystem >= 100GB: ❌ Failed
    - Reason: Root FS is only 43GB, required: 100GB

- NIC Configuration: ℹ️ INFO
    - Reason: Available network interfaces: ens3 | Speeds (Mbps): -1

- Jumbo Frames Enabled: ❌ Failed
    - Reason: MTU is 1500, recommended > 1500

- NIC Static IP Configuration: ❌ Failed
    - Reason: NIC is using DHCP, static IP is recommended

- NIC Bandwidth (10GbE Recommended): ❌ Failed
    - Reason: NIC speed is -1 Mbps, recommended is 10GbE

- Network Latency: ℹ️ INFO
    - Reason: Average latency (ms): ['0.111']

==================================================
** Summary **
--------------------------------------------------
❌ Critical Failures Detected:
   - Tuned Profile, RHEL Profile, Minimum RAM, Swap Space, /var Partition, Root Filesystem, Jumbo Frames Enabled, NIC Static IP Configuration, NIC Bandwidth

** Action Required: Please resolve these issues before proceeding.

❯ pwd                                                                                                                                                                                                                                      ─╯
/home/kushaldeb/Github/cephadm-ansible/reports

░▒▓ ~/Github/cephadm-ansible/reports  on implement_os_preflight_checks *1 

❯ ls -l                                                                                                                                                                                                                                    ─╯
total 4
-rw-r--r--. 1 kushaldeb kushaldeb 1872 Feb 24 22:04 rhel-ceph-admin_preflight_report.txt


```

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+
+            - name: Ensure OS is RHEL 9+
+              fail:
+                msg: "Ceph requires RHEL 9+. Detected: {{ansible_facts['distribution']}} {{ansible_facts['distribution_version']}}"
```
I believe ansible linter would complain:
```suggestion
                msg: "Ceph requires RHEL 9+. Detected: {{ ansible_facts['distribution'] }} {{ ansible_facts['distribution_version'] }}"
```

Kushal-deb replied:
Addressed in the latest commit.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+              fail:
+                msg: "Ceph requires RHEL 9+. Detected: {{ansible_facts['distribution']}} {{ansible_facts['distribution_version']}}"
+              when: 
+                - ansible_facts['distribution'] not in ['RedHat'] or ansible_facts['distribution_major_version'] | int <9
```
```suggestion
                - ansible_facts['distribution'] not in ['RedHat'] or ansible_facts['distribution_major_version'] | int < 9
```

Kushal-deb replied:
Addressed in the latest commit.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+              set_fact:
+                preflight_results: "{{ preflight_results + [{'Check': 'OS Version', 'Result': 'PASS' if ansible_facts['distribution_major_version'] | int >= 9 else 'FAIL'}] }}"
+            
+            - name: Check if SELinux is in Enforcing Mode
+              command: getenforce
+              register: selinux_status
+              changed_when: false
+              ignore_errors: true
```
I would strongly advise against using `ignore_errors: true`.

Wouldn't something like the following be better?

```suggestion
            - name: Check if SELinux is in Enforcing Mode
              ansible.posix.selinux:
                policy: targeted
              register: selinux_status
              changed_when: false
```

Kushal-deb replied:
Using the above suggestion to set the SELinux.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+              changed_when: false
+              ignore_errors: true
+
+            - name: Fail if SELinux is not Enforcing
```
I would rework this task and get rid of that `ignore_errors: true`

Kushal-deb replied:
I have removed `ignore_errors:true` in most of the cases.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+              package_facts:
+                manager: auto
+
+            - name: Initialize missing packages list
+              set_fact:
+                missing_packages: []
+
+            - name: Identify missing required packages
+              set_fact:
+                missing_packages: "{{ missing_packages | default([]) + [item] }}"
+              when: item not in ansible_facts.packages
+              loop:
+                - rpcbind
+                - podman
+                - firewalld
+
+            - name: Install missing required packages
+              package:
+                name: "{{ missing_packages }}"
+                state: present
+              when: missing_packages | length > 0
```
This is weird, IMHO it doesn't make sense, it goes against the principle of idempotency:
The ansible module `package` already handles that logic/complexity for you.

Kushal-deb replied:
Thanks for the suggestion. I have made the changes.

guits reviewed (changes requested):
Please, avoid as much as possible using `ignore_errors: true`

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,224 @@
+            - name: Store OS check result
+              set_fact:
+                preflight_results: "{{ preflight_results + [{'Check': 'OS Version', 'Result': 'PASS' if ansible_facts['distribution_major_version'] | int >= 9 else 'FAIL'}] }}"
+            
```
```suggestion

```

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
+              set_fact:
+                preflight_results: []
+
+            - name: Ensure OS is RHEL 9+
+              fail:
+                msg: "Ceph requires RHEL 9+. Detected: {{ ansible_facts['distribution'] }} {{ ansible_facts['distribution_version'] }}"
+              when: 
+                - ansible_facts['distribution'] not in ['RedHat'] or ansible_facts['distribution_major_version'] | int < 9
+              register: os_check
+              ignore_errors: true
```
Since when do we want to support only RHEL?

Kushal-deb replied:
The requirement was to add a check for RHEL 9+ according to the epic. 
Based on the specified criteria, I ensured that the preflight check enforces this requirement. However, if the scope needs to include other distributions, I can refine the condition accordingly. Please let me know if I should do it in a different way.
Thanks.

Epic link: https://jsw.ibm.com/browse/ISCE-1612

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
       when: ansible_facts['os_family'] == 'RedHat'
       block:
+        - name: Initialize os_check variable
+          set_fact:
+            os_check: { "failed": false }
```
I don't get why you need this fact, it seems not needed at all.
```suggestion
          set_fact:
            os_check: { "failed": false }
```

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
+              when: 
+                - ansible_facts['distribution'] not in ['RedHat'] or ansible_facts['distribution_major_version'] | int < 9
+              register: os_check
+              ignore_errors: true
```
please, get rid of this `ignore_errors: true`
```suggestion
```

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
+                path: templates/preflight_report.j2
+              register: template_stat
+
+            - name: Create the preflight report template if missing
+              delegate_to: localhost
+              run_once: true
+              become: false
+              copy:
+                dest: templates/preflight_report.j2
+                content: |
+                  Preflight Check Report
+                  ======================
+                  {% for item in preflight_results %}
+                  - {{ item['Check'] }}: {{ item['Result'] }}
+                    {% if item['Result'] == 'FAIL' and item['Reason'] != 'N/A' %}
+                    -> Reason: {{ item['Reason'] }}
+                    {% endif %}
+                  {% endfor %}
+              when: not template_stat.stat.exists
```
Not a big fan of this approach, we should rather maintain that template within the repo

Kushal-deb replied:
Created a static file and using that to generate the report.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
+                  }
+                ] }}"
+
+            - name: Ensure required packages are installed
+              package:
+                name:
+                  - rpcbind
+                  - podman
+                  - firewalld
+                state: present
```
```suggestion
            - name: Ensure required packages are installed
              package:
                name:
                  - rpcbind
                  - podman
                  - firewalld
                state: present
```

we have a variable `infra_pkgs` in `ceph_defaults` for this

Kushal-deb replied:
Using infra_pkgs in the current implementation.

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,205 @@
+                msg: "Podman is installed at: {{ podman_path}}"
+
+            - name: Extract Podman version
+              shell: "podman --version | awk '{print $3}'"
```
not a big fan of this approach, output scrapping like this is not ideal:
For instance, it wouldn't be the first time I see a CLI returning a new format after a release which would break this.

Kushal-deb replied:
Using `package_facts` for package detection and `ansible_facts.packages `for version extraction

guits commented on cephadm-preflight.yml:
```diff
@@ -48,6 +48,144 @@
+              set_fact:
+                preflight_results: []
+
+            - name: Ensure OS is RHEL 9+
+              fail:
+                msg: "Ceph requires RHEL 9+. Detected: {{ ansible_facts['distribution'] }} {{ ansible_facts['distribution_version'] }}"
+              when: 
+                - ansible_facts['distribution'] not in ['RedHat'] or ansible_facts['distribution_major_version'] | int < 9
```
This task breaks CentOS Stream support

Kushal-deb replied:
Moved the implementation to different playbooks for now.

Would it still hit the issue above?

Kushal-deb replied:
The task is now updated to fail only when the OS is not RHEL 9+. Tested with RedHat 9.5, the condition evaluates to false, and the task is correctly skipped. This ensures that the check does not unnecessarily interrupt execution on supported systems. Please let me know if you have any concerns!

Thanks.

guits commented on tasks/preflight_cpu_mem_checks.yml:
```diff
@@ -0,0 +1,118 @@
+    ] }}"
+
+- name: Check available CPU cores
+  shell: "nproc"
```
no need to use a `shell` task here
```suggestion
  ansible.builtin.command: "nproc"
```

guits replied:
by the way, that's probably something you want to get from ansible facts :

```
        "processor": [
            "0",
            "AuthenticAMD",
            "AMD EPYC-Rome Processor",                                                                                                                                                                                                                                                                                                                                                                                                                  "1",
            "AuthenticAMD",
            "AMD EPYC-Rome Processor",
            "2",
            "AuthenticAMD",
            "AMD EPYC-Rome Processor",
            "3",
            "AuthenticAMD",
            "AMD EPYC-Rome Processor"
        ],
        "processor_cores": 1,
        "processor_count": 4,
        "processor_nproc": 4,
        "processor_threads_per_core": 1,
        "processor_vcpus": 4,
```

Kushal-deb replied:
Addressed

guits commented on tasks/preflight_cpu_mem_checks.yml:
```diff
@@ -0,0 +1,118 @@
+      }
+    ] }}"
+
+- name: Check total RAM available
+  shell: "free -m | awk '/Mem:/ {print $2}'"
+  register: total_ram
+  changed_when: false
+  failed_when: false
```
ditto

```
        "memfree_mb": 1104,
        "memory_mb": {
            "nocache": {
                "free": 2927,
                "used": 731
            },
            "real": {
                "free": 1104,
                "total": 3658,
                "used": 2554
            },
            "swap": {
                "cached": 7,
                "free": 1903,
                "total": 2047,
                "used": 144
            }
        },
        "memtotal_mb": 3658,
```

Kushal-deb replied:
Addressed

guits commented on tasks/preflight_cpu_mem_checks.yml:
```diff
@@ -0,0 +1,118 @@
+      }
+    ] }}"
+
+- name: Check total swap available
+  shell: "free -m | awk '/Swap:/ {print $2}'"
+  register: total_swap
+  changed_when: false
+  failed_when: false
```
ditto

```
        "swapfree_mb": 1903,
        "swaptotal_mb": 2047,
```

Kushal-deb replied:
Addressed

guits commented on tasks/preflight_cpu_mem_checks.yml:
```diff
@@ -0,0 +1,118 @@
+      }
+    ] }}"
+        
+- name: Check if /var is a separate partition
+  shell: "df -hT | grep '/var'"
+  register: var_partition
+  changed_when: false
+  failed_when: false
+
+- name: Check root filesystem size
+  shell: "df -hT | grep ' /$' | awk '{print $3}'"
+  register: root_fs_size
+  changed_when: false
+  failed_when: false
```
`ansible_facts['mounts']` exposes a list with all mounts

Kushal-deb replied:
Addressed.

Kushal-deb committed:
Add preflight OS and NIC and other checks

- Implemented OS preflight checks to validate system requirements before Ceph cluster creation.
- Checks include:
  - OS version (RHEL 9+ required)
  - SELinux enforcing mode
  - Firewalld installation and status
  - Required package availability (rpcbind, podman, firewalld)
  - Podman version check (>= 3.3)
  - RHEL software profile validation
  - Tuned profile check
  - CPU, RAM, Swap, and Filesystem (part of other checks)
  - Check whether jumbo frames are enabled
  - Is it configured with DHCP or static IP
  - Is the bandwidth sufficient
  - Collect and output current NIC options set (e.g. Bonding, not bridged or virtual)
  - Check and report network latency (ping) with all hosts provided in the inventory file
  - Listing all NICs

Signed-off-by: Kushal Deb <Kushal.Deb@ibm.com>

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,360 @@
+    src: templates/preflight_report.j2
+    dest: ./ceph_preflight_report.txt
+
+- name: Display Preflight Check Report
+  command: cat ./ceph_preflight_report.txt
+  register: report_output
+  changed_when: false
```
I think you can get rid of this task
```suggestion
```

you probably want to add a `register: report_output` to the task "Generate Preflight Check Report" above

Kushal-deb replied:
I tried adding the register `report_output` to that task but while displaying the result, it could only output the metadata and not the actual report.
Tested with this below task:

```
- name: Generate Preflight Check Report
  delegate_to: localhost
  run_once: true
  become: false
  template:
    src: templates/preflight_report.j2
    dest: ./ceph_preflight_report.txt
  register: report_output  

- name: Debug Report Output
  debug:
    var: report_output
```

```
TASK [Debug Report Output] ************************************************************************************************************************************
ok: [rhel-ceph-admin] => {
    "report_output": {
        "changed": true,
        "checksum": "22398d99cac1638f0c101608967f5d547a643a69",
        "dest": "./ceph_preflight_report.txt",
        "failed": false,
        "gid": 1001,
        "group": "kushaldeb",
        "md5sum": "5152ba2d7bf839cd4c3e57e396402487",
        "mode": "0644",
        "owner": "kushaldeb",
        "size": 2152,
        "src": "/home/kushaldeb/.ansible/",
        "state": "file",
        "uid": 1000
    }
}
```

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,360 @@
+- name: Ping all hosts in inventory
+  ansible.builtin.ping:
+  register: ping_results
+  ignore_errors: yes
```
```suggestion
  failed_when: false
```

Kushal-deb replied:
Addressed

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,360 @@
+    preflight_results: "{{ preflight_results + [{'Check': 'CPU Cores >= 4', 'Result': cpu_core_check, 'Reason': cpu_core_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['CPU Cores'] if cpu_core_check == 'FAIL' else preflight_failures }}"
+
+- name: Get total RAM using Ansible facts
+  set_fact:
+    total_ram: "{{ ansible_facts['memtotal_mb'] | int }}"
```
why do you need to set a fact for this ?

Kushal-deb replied:
Removed it.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,360 @@
+    preflight_results: "{{ preflight_results + [{'Check': 'Root Filesystem >= 100GB', 'Result': root_fs_check, 'Reason': root_fs_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['Root Filesystem'] if root_fs_check == 'FAIL' else preflight_failures }}"
+
+- name: Get active network interfaces
+  ansible.builtin.command: "ip -j addr show"
+  register: ip_info
+  changed_when: false
+  failed_when: false
```
doesn't `ansible_facts` provide this information...?

Kushal-deb replied:
Addressed.

asm0deuz commented:
jenkins test el9-functional

Kushal-deb commented:
jenkins test el9-functional

asm0deuz commented:
jenkins test el9-functional

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+---
```
please, make it an actual playbook:

```suggestion
---
- hosts: all
  become: true
  gather_facts: true
  tasks:
    - name: Initialize preflight results list
      set_fact:
        preflight_results: []
        preflight_failures: []
```

Kushal-deb replied:
Updated it to be an actual playbook.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+    preflight_results: "{{ preflight_results + [{'Check': 'SELinux', 'Result': selinux_check, 'Reason': selinux_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['SELinux'] if selinux_check == 'FAIL' else preflight_failures }}"
+
+- name: Ensure required packages are installed
+  package:
+    name: "{{ infra_pkgs }}"
+    state: present
+  register: package_install
+  failed_when: false
+
+- name: Determine Package Installation Check Result
+  set_fact:
+    package_check: "{{ 'PASS' if not package_install.failed else 'FAIL' }}"
+
+- name: Determine Package Installation Failure Reason
+  set_fact:
+    package_reason: "{{ 'Some required packages failed to install' if package_check == 'FAIL' else 'N/A' }}"
+
+- name: Store Package Installation Result
+  set_fact:
+    preflight_results: "{{ preflight_results + [{'Check': 'Required Packages Installed', 'Result': package_check, 'Reason': package_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['Required Packages'] if package_check == 'FAIL' else preflight_failures }}"
```
`cephadm-preflight` playbook already handles package installation.
```suggestion
```

Kushal-deb replied:
Removed this part. 
Just storing if the packages are installed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+    preflight_results: "{{ preflight_results + [{'Check': 'Required Packages Installed', 'Result': package_check, 'Reason': package_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['Required Packages'] if package_check == 'FAIL' else preflight_failures }}"
+
+- name: Ensure firewalld is enabled and running
+  systemd:
+    name: firewalld
+    state: started
+    enabled: true
+  register: firewall_status
+  failed_when: false
```
We should just move this task to `cephadm-preflight.yml`
```suggestion
```

Kushal-deb replied:
Moved it to cephadm-preflight.yml

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  set_fact:
+    preflight_results: "{{ preflight_results + [{'Check': 'Podman Installed', 'Result': podman_check, 'Reason': podman_reason}] }}"
+
+- name: Ensure Podman is installed if missing (Fixable)
+  package:
+    name: podman
+    state: present
+  when: not podman_installed
```
`cephadm-preflight.yml` already handles podman installation
```suggestion
```

Kushal-deb replied:
Removed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  set_fact:
+    preflight_results: "{{ preflight_results + [{'Check': 'NIC Bandwidth (10GbE Recommended)', 'Result': nic_speed_check, 'Reason': nic_speed_reason}] }}"
+
+- name: Extract NIC Details
+  set_fact:
+    nic_config_details: "{{ ansible_facts['interfaces'] }}"
```
This task is unnecessary, just use `ansible_facts['interfaces']` directly wherever you need it
```suggestion
```

Kushal-deb replied:
Addressed

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  set_fact:
+    preflight_results: "{{ preflight_results + [{'Check': 'NIC Configuration', 'Result': 'INFO', 'Reason': 'NIC options: ' ~ nic_config_details | join(', ')}] }}"
+
+- name: Identify Front-End and Back-End NICs
+  set_fact:
+    frontend_nic: "{{ ansible_facts['default_ipv4']['interface'] | default('Unknown') }}"
+    backend_nic: "{{ ansible_facts['interfaces'] | difference(['lo', ansible_facts['default_ipv4']['interface']]) | first | default(ansible_facts['default_ipv4']['interface']) }}"
```
This task feels a bit awkward; that's not something Ansible can predict for you
```suggestion
```
I'd simply list all available network interfaces along with their link speed 🤷‍♂️

Kushal-deb replied:
Listing the available NICs

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+    src: templates/preflight_report.j2
+    dest: ./ceph_preflight_report.txt
+
+- name: Read Preflight Check Report
+  slurp:
+    src: ./ceph_preflight_report.txt
+  register: report_content
```
Is this task really needed? By the way, it's probably missing a `delegate_to: localhost` so I suspect it would fail.

please, have a look at `loopkup('template')` instead

Kushal-deb replied:
Removed the task.
Using a debug to display the report instead.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+---
```
nit: I'd keep consistency and rename that file : `preflight-checks.yml` ( `_` -> `-` )

Kushal-deb replied:
Addressed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+    preflight_results: []
+    preflight_failures: []
+
+- name: Gather all Ansible facts
+  setup:
```
You can get rid of the following if you make it an actual playbook with `gather_facts: true` 

```suggestion
```

Kushal-deb replied:
Addressed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+
+- name: Store Firewalld check result
+  set_fact:
+    firewalld_check: "{{ 'PASS' if firewall_status.status.ActiveState == 'active' else 'FAIL' }}"
+    firewalld_reason: "{{ 'Firewalld was not running and could not be started' if firewall_status.failed else 'N/A' }}"
```
duplicate?
```suggestion
```

Kushal-deb replied:
Addressed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+
+- name: Store SELinux check result
+  set_fact:
+    selinux_check: "{{ 'PASS' if ansible_facts['selinux']['status'] == 'enabled' and ansible_facts['selinux']['mode'] == 'enforcing' else 'FAIL' }}"
+    selinux_reason: "{{ 'SELinux was not in enforcing mode and could not be enforced automatically' if selinux_check == 'FAIL' else 'N/A' }}"
```
duplicate?
```suggestion
```

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+    preflight_results: "{{ preflight_results + [{'Check': 'Firewalld Running', 'Result': firewalld_check, 'Reason': firewalld_reason}] }}"
+    preflight_failures: "{{ preflight_failures + ['Firewalld Running'] if firewalld_check == 'FAIL' else preflight_failures }}"
+
+- name: Collect installed package facts
+  package_facts:
+    manager: auto
```
I would move this task to the beginning of the playbook

Kushal-deb replied:
- Required Packages Installed: 
      ✅ Passed

guits replied:
@Kushal-deb it looks like you removed this task, this causes an error:

```
The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'packages'. 'dict object' has no attribute 'packages'
```

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  changed_when: false
+  failed_when: selinux_status.failed
+
+- name: Retrieve SELinux status from ansible_facts
+  setup:
+    gather_subset:
+      - selinux
```
```suggestion
```

Kushal-deb replied:
Addressed

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  set_fact:
+    preflight_results: "{{ preflight_results + [{'Check': 'Separate NICs for Frontend & Backend Networks', 'Result': nic_separation_check, 'Reason': nic_separation_reason}] }}"
+
+- name: Ping all hosts in inventory
+  ansible.builtin.ping:
+  register: ping_results
+  failed_when: false
+  delegate_to: "{{ item }}"
+  with_items: "{{ groups['all'] }}"
```
This won't give you the latency like an actual ICMP ping would.
This is just an Ansible module called 'ping', used to check if Ansible is able to connect to nodes.

Kushal-deb replied:
Changed it to:
```
  - name: Ping all hosts in inventory to measure latency
      ansible.builtin.shell: "ping -c 4 {{ item }} | grep 'rtt min/avg/max/mdev' | awk -F'/' '{print $5}'"
      register: ping_results
      changed_when: false
      failed_when: false
      delegate_to: "{{ item }}"
      with_items: "{{ groups['all'] }}"
```

guits commented on cephadm-preflight.yml:
```diff
@@ -45,6 +46,9 @@
       import_role:
         name: ceph_defaults
 
+    - name: Run checks
+      import_tasks: preflight_checks.yml
```
```suggestion
    - name: variables validations
       ansible.builtin.import_playbook: validate/preflight.yml
```

Kushal-deb replied:
Addressed.

guits commented on preflight_checks.yml:
```diff
@@ -0,0 +1,314 @@
+  setup:
+
+- name: Check if OS is RHEL 9+
+  set_fact:
```
please, use FQCN everywhere:
```suggestion
  ansible.builtin.set_fact:
```

Kushal-deb replied:
Updated with FQCN.

guits commented:
@Kushal-deb Please consider using a single `set_fact` task, like the following, instead of multiple `set_fact` tasks for each check:

```yaml
    - name: Store all check results
      set_fact:
        preflight_results: >-
          {{ preflight_results + [
            {'Check': 'OS Version', 'Result': os_check, 'Reason': os_reason},
            {'Check': 'Tuned Profile', 'Result': tuned_profile_check, 'Reason': tuned_profile_reason},
            {'Check': 'RHEL Profile', 'Result': rhel_profile_check, 'Reason': rhel_profile_reason},
            {'Check': 'Firewalld Running', 'Result': firewalld_check, 'Reason': firewalld_reason},
            {'Check': 'Podman Installed', 'Result': podman_check, 'Reason': podman_reason},
            {'Check': 'SELinux', 'Result': selinux_check, 'Reason': selinux_reason},
            {'Check': 'Minimum RAM (8GB)', 'Result': memory_checks['ram']['result'], 'Reason': memory_checks['ram']['reason']},
            {'Check': 'Swap Space (1.5x RAM)', 'Result': memory_checks['swap']['result'], 'Reason': memory_checks['swap']['reason']},
            {'Check': 'CPU x86-64-v2', 'Result': cpu_checks['x86_64_v2']['result'], 'Reason': cpu_checks['x86_64_v2']['reason']},
            {'Check': 'CPU Cores >= 4', 'Result': cpu_checks['cores']['result'], 'Reason': cpu_checks['cores']['reason']},
            {'Check': '/var is a separate partition', 'Result': filesystem_checks['var_partition']['result'], 'Reason': filesystem_checks['var_partition']['reason']},
            {'Check': 'Root Filesystem >= 100GB', 'Result': filesystem_checks['root_fs']['result'], 'Reason': filesystem_checks['root_fs']['reason']},
            {'Check': 'SELinux', 'Result': selinux_check, 'Reason': selinux_reason},
            {'Check': 'Jumbo Frames Enabled', 'Result': jumbo_frames_check, 'Reason': jumbo_frames_reason},
            {'Check': 'Network Latency', 'Result': 'INFO', 'Reason': 'Latency results: ' ~ ping_results.results | map(attribute='ping') | list},
            {'Check': 'NIC Static IP Configuration', 'Result': nic_config_check, 'Reason': nic_config_reason},
            {'Check': 'NIC Bandwidth (10GbE Recommended)', 'Result': nic_speed_check, 'Reason': nic_speed_reason},

          ] }}
        preflight_failures: >-
          {{ preflight_failures
             + (['OS Version'] if os_check == 'FAIL' else [])
             + (['Tuned Profile'] if tuned_profile_check == 'FAIL' else [])
             + (['RHEL Profile'] if rhel_profile_check == 'FAIL' else [])
             + (['SELinux'] if selinux_check == 'FAIL' else [])
             + (['Firewalld Running'] if firewalld_check == 'FAIL' else [])
             + (['Podman Installed'] if not podman_installed else [])
             + (['Minimum RAM'] if memory_checks['ram']['result'] == 'FAIL' else [])
             + (['Swap Space'] if memory_checks['swap']['result'] == 'FAIL' else [])
             + (['CPU x86-64-v2'] if cpu_checks['x86_64_v2']['result'] == 'FAIL' else [])
             + (['CPU Cores'] if cpu_checks['cores']['result'] == 'FAIL' else [])
             + (['/var Partition'] if filesystem_checks['var_partition']['result'] == 'FAIL' else [])
             + (['Root Filesystem'] if filesystem_checks['root_fs']['result'] == 'FAIL' else [])
             + (['SELinux'] if selinux_check == 'FAIL' else []) }}
```

guits commented on preflight-checks.yml:
```diff
@@ -0,0 +1,205 @@
+          (['NIC Bandwidth'] if primary_speed | int < 10000 else []) 
+        }}"
+
+    - name: Generate preflight check report file
+      ansible.builtin.template:
+        src: preflight_report.j2
+        dest: ./preflight_report.txt
+      delegate_to: localhost
+      run_once: true
+      become: false
+
+    - name: Show Preflight Check Report
+      ansible.builtin.debug:
+        msg: "{{ lookup('template', 'preflight_report.j2') | split('\n') }}"
```
Do we actually need to generate the report on disk? If so, you can ignore this comment. Otherwise, you could do something like the following:

```suggestion
    - name: Generate Preflight Check Report in memory
      set_fact:
        report_content: "{{ lookup('template', 'templates/preflight_report.j2') }}"

    - name: Show Report Summary
      debug:
        msg: "{{ report_content }}"
```

Kushal-deb replied:
Thanks for the suggestion! The report file serves as a reference for users who may need to audit the preflight checks later. Since this playbook is often used in environments where logs and reports are required for troubleshooting, having a persistent report file makes it easier to track issues.

However, I agree that generating the report in memory can be useful for quick debugging. If needed, we could add a conditional flag (save_report: true/false) to allow users to choose whether they want to generate the file. Let me know if that sounds reasonable!

guits reviewed (changes requested):
@Kushal-deb what happens if you have more than 1 node in your inventory host ?
At the end of the playbook, the report seems to be generated only once, for the first node only.

guits commented:
jenkins test el9-functional

Kushal-deb commented:
> @Kushal-deb what happens if you have more than 1 node in your inventory host ? At the end of the playbook, the report seems to be generated only once, for the first node only.

Hi, I have updated the implementation to generate the result files on the controller node under reports directory.

example:

```
░▒▓ ~/Github/cephadm-ansible/reports  on implement_os_preflight_checks *1 ?8 ▓▒░·······································································································································░▒▓ at 22:09:10 ▓▒░─╮
❯ pwd                                                                                                                                                                                                                                      ─╯
/home/kushaldeb/Github/cephadm-ansible/reports

░▒▓ ~/Github/cephadm-ansible/reports  on implement_os_preflight_checks *1 ?8 ▓▒░·······································································································································░▒▓ at 22:09:11 ▓▒░─╮
❯ ls -l                                                                                                                                                                                                                                    ─╯
total 4
-rw-r--r--. 1 kushaldeb kushaldeb 1872 Feb 24 22:04 rhel-ceph-admin_preflight_report.txt
```
